Implement multiple lights and more Light types

ffreyer opened the pull request:
# Description

## Change Summary

This pr adds/changes:
- (breaking) `backlight` now affects `normal` instead of `lightdir`. This means specular reflections show up with backlight
- (breaking) changed default light direction (position), light intensities, diffuse and specular attributes 
- (minor breaking) change default light to `DirectionalLight` (closes #3239)
- (minor breaking) added smoothing of `diff_coeff` around 0. This softens the edge where normals are perpendicular to the light direction)
- (non-breaking change) `shading` attribute is now `:none/:fast/:verbose` instead of true or false
- (internal/maybe breaking) moved scaling in Axis3 from view to model matrix and stopped connecting scenes/plots with different spaces. The latter is something I want to do for #3013.
- (new) support for multiple light sources (need not be unique, ref #1592)
- (new) `PointLight` now has attenuation parameters (distance based intensity falloff, off by default)
- (new) added `DirectionalLight` (parallel light rays)
- (new) added `SpotLight` (light source illuminating a cone)
- (fix) the light color of `PointLight` (and all others) is now connected (fixes #3238)
- (internal) `colorbuffer()` now polls events to trigger `render_tick` so that it can cause lighting updates
- (internal) light calculation moved to world space which shouldn't change anything visually
- (internal) loosen type restriction on `lights` vector (fix #2985)


## Type of change

- [x] breaking (backlight, default light, smoothing around 0 intensity)

## TODO

- [x] Hook up updates (for individual lights and the lights array)
- [x] implement more simple light types (SpotLight, DirectionalLight, PointLight with attentuation)
- [x] Hook up `shading`
- [x] Update other backends to work with changes made to the lighting interface
- [x] Implement new lights in RPRMakie
- [x] lighting in volume 
- [x] backlight
- [x] record updates
- [x] Added an entry in NEWS.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [x] Added unit tests for new algorithms, conversion methods, etc.
- [x] Added reference image tests for new plotting functions, recipes, visual options, etc.
- [x] check/fix missing normals for matcap with NoShading
- [x] fix WGLMakie
- [x] fix laggy directional light in WGLMakie (move camera-space light direction calculation to javascript)
- [x] fix Axis3
- [x] fix images in docs (MultiLightShading and SSAO fail to render correctly)

ffreyer commented:
I'm following the attenuation from learnopengl/ogre3d for now, which is effectively `1 / (1 + a * distance + b * distance^2)`. Example:
```julia
fig = Figure(backgroundcolor = :black)
sl = Slider(fig[1, 1], range = 0.05:0.05:5.0)

lights = Makie.AbstractLight[
    AmbientLight(RGBf(0.1, 0.1, 0.1)),
    PointLight(RGBf(1,0,0), Point3f(2, 2, 0), 1.0),
    PointLight(RGBf(0,1,0), Point3f(-2, 2, 0), 10.0),
    PointLight(RGBf(0,0,1), Point3f(-2, -2, 0), 100.0),
    PointLight(RGBf(1,1,1), Point3f(2, -2, 0), 10.0),
]
scene = LScene(
    fig[2, 1], show_axis = false,
    scenekw=(camera = cam3d!, lights = lights, backgroundcolor = :black, center = false),
)

p = mesh!(scene, Rect3f(Point3f(-10, -10, 0.01), Vec3f(20, 20, 0.02)), color = :white)
map(z -> translate!(p, 0, 0, -z), sl.value)

update_cam!(scene.scene, Vec3f(0, 0, 10), Vec3f(0, 0, 0), Vec3f(0, 1, 0))

idxs = vcat(1:length(sl.range[]), length(sl.range[])-1:-1:2)
record(fig, "attenuation.webm", idxs, framerate = 30) do idx
    sl.selected_index[] = idx
end
```
[attenuation.webm](https://github.com/MakieOrg/Makie.jl/assets/10947937/ebfff235-1823-4038-91b7-308be110ca09)

I also tried just `a / distance^3` as an attenuation factor beforehand, with a weak directional blue light above:
```julia
lights = [
    AmbientLight(RGBf(0.1, 0.1, 0.1)),
    PointLight(RGBf(0,1,0), Point3f(0, -3, 1)),
    PointLight(RGBf(1,0,0), Point3f(-3, 0, 1)),
    PointLight(RGBf(0,1,0), Point3f(0,  3, 1)),
    PointLight(RGBf(1,0,0), Point3f( 3, 0, 1)),
    Makie.DirectionalLight(RGBf(0, 0, 0.5), Vec3f(0, 0, 1)),
]
scene = Scene(camera = cam3d!, lights = lights)
mesh!(scene, Rect3f(Point3f(-1), Vec3f(2)), color = :white)
r = range(-10, 10, length=101)
surface!(scene, r, r, [0.5 * (sin(x+y) * sin(2x) * sin(3y)) - 1 for x in r, y in r], colormap = [:white, :white])
ps = [light.position[] for light in lights if light isa PointLight]
scatter!(scene, ps, color = :yellow, strokewidth = 1, strokecolor = :black)

screen = display(scene)
```
![Screenshot from 2023-09-21 18-46-22](https://github.com/MakieOrg/Makie.jl/assets/10947937/93acb0f8-565b-42eb-86b4-21b135faa545)

ffreyer commented on GLMakie/src/drawing_primitives.jl:
```diff
@@ -2,6 +2,75 @@ using Makie: transform_func_obs, apply_transform
+    max_lights = 64
+    max_params = 5 * 64
+
+    updater = screen.render_tick
+    MAX_PRIORITY = typemax(Int)
+
+    attr[:lights_length] = map(updater, priority = -1000) do _
```
Updating on `render_tick` doesn't work with `record`.

Updating (many) lights is kind of complicated in general, because we want to react to updates of each field for each light and also when lights are added or deleted to the lights vector.

Listening to each field individually would require attaching and removing listeners when the lights vector changes. Detecting that would either require some kind of Vector wrapper or an update on draw.

Instead of bothering with all of that I thought it would be easier to just check the lights vector once per frame and update it then. I'm using `render_tick` for that atm, but I guess we don't want that to update in `record` to avoid triggering events?

ffreyer commented:
<details>

<summary>Moving, colored `PointLight`s with different attenuation + white SpotLight.</summary>

```julia
begin
    fig = Figure(backgroundcolor = :black)
    sl = Slider(fig[1, 1], range = 0.05:0.05:5.0)

    # Prepare lights
    angle = Observable(0)
    angle2pos(phi) = Point3f(3cosd(phi), 3sind(phi), 0)
    lights = Makie.AbstractLight[
        AmbientLight(RGBf(0.1, 0.1, 0.1)),
        PointLight(RGBf(1,0,0), map(phi -> angle2pos(phi + 0), angle), 10.0),
        PointLight(RGBf(0,1,0), map(phi -> angle2pos(phi + 120), angle), 100.0),
        PointLight(RGBf(0,0,1), map(phi -> angle2pos(phi + 240), angle), 1000.0),
        SpotLight(RGBf(1,1,1), Point3f(0, 0, 1), Vec3f(0, 0, -1), pi/6),
    ]

    # Set scene
    scene = LScene(
        fig[2, 1], show_axis = false,
        scenekw=(camera = cam3d!, lights = lights, backgroundcolor = :black, center = false),
    )

    # floor
    p = mesh!(scene, Rect3f(Point3f(-10, -10, 0.01), Vec3f(20, 20, 0.02)), color = :white)
    map(z -> translate!(p, 0, 0, -z), sl.value)

    # Random balls
    N_balls = 2_000
    random_colors = !true
    meshscatter!(
        scene, [Point3f(20rand()-10, 20rand()-10, 10rand()-5) for _ in 1:N_balls],
        color = random_colors ? rand(RGBf, N_balls) : (:white)
    )

    # Visualize light source positions
    ps = Observable([light.position[] for light in lights if light isa Union{PointLight, SpotLight}])
    cs = [light.color[]    for light in lights if light isa Union{PointLight, SpotLight}]
    scatter!(scene, ps, color = cs, strokewidth = 1, strokecolor = :white)

    # Place camera above
    # update_cam!(scene.scene, Vec3f(0, 0, 10), Vec3f(0, 0, 0), Vec3f(0, 1, 0))

    # place camera at angle nearby
    update_cam!(scene.scene, Vec3f(8, 1, 4), Vec3f(0, 0, 0), Vec3f(0, 0, 1))

    # record
    idxs = vcat(1:length(sl.range[]), length(sl.range[])-1:-1:2)
    record(fig, "point_lights_and_spot_light.webm", idxs, framerate = 30) do idx
        sl.selected_index[] = idx
        angle[] = angle[] + 4
        ps[] = [light.position[] for light in lights if light isa Union{PointLight, SpotLight}]
        notify(fig.scene.current_screens[1].render_tick)
    end

    # interactive
    # display(fig)
    # @async while isopen(fig.scene)
    #     angle[] = angle[] + 2
    #     ps[] = [light.position[] for light in lights if light isa Union{PointLight, SpotLight}]
    #     sleep(1/60)
    # end
end
```

</details>

[point_lights_and_spot_light.webm](https://github.com/MakieOrg/Makie.jl/assets/10947937/80635fe4-d7ee-4dec-81b6-1cc3545bc80b)

ffreyer commented:
Added refimgs:

Ambient + SpotLights + PointLights:
![Screenshot from 2023-09-26 14-57-39](https://github.com/MakieOrg/Makie.jl/assets/10947937/ddc90b2f-701b-4682-8dc4-32fee75f1184)

DirectionalLights:
![Screenshot from 2023-09-26 14-57-46](https://github.com/MakieOrg/Makie.jl/assets/10947937/d12b9a3f-85ab-4e7d-9831-ce2730d2dcff)

ffreyer commented on GLMakie/src/screen.jl:
```diff
@@ -686,7 +687,7 @@ function Makie.colorbuffer(screen::Screen, format::Makie.ImageStorageFormat = Ma
     ctex = screen.framebuffer.buffers[:color]
     # polling may change window size, when its bigger than monitor!
     # we still need to poll though, to get all the newest events!
-    # GLFW.PollEvents()
+    pollevents(screen)
     # keep current buffer size to allows larger-than-window renders
     render_frame(screen, resize_buffers=false) # let it render
```
I added `pollevents()` here so lighting updates from `render_tick` can take effect before drawing the frame. Given the note I would assume this breaks output sizes with `save` and `record`, however I have no problems saving/recording at `resolution = (4000, 4000)`.

ffreyer commented on GLMakie/assets/shader/util.vert:
```diff
@@ -239,30 +239,41 @@ out vec3 o_view_pos;
-    // direction to light
-    o_lightdir = normalize(view*vec4(lightposition, 1.0) - view_pos).xyz;
+
+    // direction to light for :fast shading
+    // (normalizing this will result in incorrect lighting)
+    vec4 view_light_pos = view*vec4(lightposition, 1.0);
+    o_lightdir = view_light_pos.xyz / view_light_pos.w - view_pos.xyz;
```
As noted normalizing here results in incorrect light directions/perceived positions. To illustrate this, consider a light source shining on a rect viewed from the side:

![Screenshot from 2023-09-27 16-27-14](https://github.com/MakieOrg/Makie.jl/assets/10947937/7739c765-f5a4-4154-8cf7-6e7fa2e6c7ac)

If we **don't** normalize `o_lightdir` it will be a vector pointing from the light source to the vertex handled by the vertex shader. Since I placed the light source at (0, 0) here they'll match p1 and p2 in the example. On the way to the fragment shader `o_lightdir` will then get interpolated between different vertices. For example, halfway between the two vertices the vector will point to the  green "Vertex center". Normalizing in the fragment shader will then give correct light-to-vertex-center direction.

If we **do** normalize `o_lightdir` in the vertex shader as before, we will be interpolating the normalized `o_lightdir` vectors. In the sketch above those point to n1 (equal to p1) and n2. Doing the same half-way interpolation for vertices as above, we end up with a `o_lightdir` halfway between n1 and n2, which is marked as "Normalized center". This is not a vector passing through the vertex center position and thus the wrong light direction.

I'm pretty sure the test failure in `Order Independent transparency` is a result of this change. There are likely quite a few more failures that didn't get caught.

ffreyer replied:
I switched to doing lighting in world space so this is not really relevant to lightdir anymore. Though it probably still applies to camdir. (World vs view/camera space shouldn't matter for lighting. It was using view/camera space because SSAO needed it and lighting could reuse it. Merging lighting between volume and the rest of Makie requires world space though.)

ffreyer commented:
Volume with multiple lights:

[volume.webm](https://github.com/MakieOrg/Makie.jl/assets/10947937/13fd974b-bc7b-4b50-8bfd-69d309baa027)

Volume plots don't differentiate inwards facing vs outwards facing surfaces, so lighting is a bit weird here. I'm not going to try to change this here though.

ffreyer commented:
Another example: 

<details>

<summary>
Meshscatter with `backlight = 1f0` synchronized with some PointLights moving through a transparent tube
</summary>

```julia
using LinearAlgebra, GeometryBasics

function tube(points, r, segments, closed = false)
    ps = sizehint!(Point3f[], segments * length(points))

    for i in eachindex(points)
        p = points[i]

        forward = Vec3f(0)
        if closed
            im = i > 1 ? i-1 : length(points)
            ip = i < length(points) ? i+1 : 1
            forward += normalize(p - points[im])
            forward += normalize(points[ip] - p)
        else
            if i > 1
                forward += normalize(p - points[i-1])
            end
            if i < length(points)
                forward += normalize(points[i+1] - p)
            end
        end
        forward = normalize(forward)

        # assuming dir != u_z
        right = cross(forward, Vec3f(0,0,1))
        up = cross(right, forward)

        # gen points
        for phi in range(0, 2pi, length = segments+1)[1:end-1]
            push!(ps, p + r * (cos(phi) * right + sin(phi) * up))
        end
    end

    fs = QuadFace[]
    for i in 0 : segments : segments * (length(points) - 2)
        for j in 1:segments
            f = QuadFace(
                i + j,
                i + mod1(j+1, segments),
                i + mod1(j+1, segments) + segments,
                i + j + segments,
            )
            push!(fs, f)
        end
    end
    if closed
        i = segments * (length(points) - 1)
        for j in 1:segments
            f = QuadFace(
                i + j,
                i + mod1(j+1, segments),
                mod1(j+1, segments),
                j,
            )
            push!(fs, f)
        end
    end

    return normal_mesh(ps, fs)
end

begin
    ps = [Point3f(cos(phi), sin(2phi), 0.2sin(phi)) for phi in range(0, 2pi, length = 101)[1:100]]
    m = tube(ps, 0.1, 8, true)

    idx = Observable(1)
    lights = Makie.AbstractLight[
        AmbientLight(RGBf(0.1, 0.1, 0.1)),
        PointLight(RGBf(1, 0, 0),   map(i -> ps[i], idx), 1.0),
        PointLight(RGBf(1, 1, 0),   map(i -> ps[mod1(i+10, length(ps))], idx), 1.0),
        PointLight(RGBf(0, 0.2, 1), map(i -> ps[mod1(i+4,  length(ps))], idx), 1.0),
        PointLight(RGBf(0, 0.8, 0), map(i -> ps[mod1(i+37, length(ps))], idx), 1.0),
        PointLight(RGBf(0, 0.8, 1), map(i -> ps[mod1(i+88, length(ps))], idx), 1.0),
    ]

    fig = Figure(backgroundcolor = :black)
    ax = LScene(fig[1, 1], scenekw = (lights = lights, camera = cam3d!), show_axis = false)
    mesh!(ax, m, color = (:white, 0.4), transparency = true, shading = :verbose)

    pos = map(idx) do i
        Point3f[
            ps[i],
            ps[mod1(i+10, length(ps))],
            ps[mod1(i+4,  length(ps))],
            ps[mod1(i+37, length(ps))],
            ps[mod1(i+88, length(ps))],
        ]
    end
    meshscatter!(ax, pos, markersize = 0.05, shading = :verbose, backlight = 1f0, color = :white) # , marker = Rect3f(Point3f(-1), Vec3f(1)))

    # display(fig)
    # @async while isopen(fig.scene)
    #     idx[] = mod1(idx[] + 1, 100)
    #     sleep(1/30)
    # end

    idxs = 1:100
    record(fig, "race.webm", idxs, framerate = 30) do _
        idx[] = mod1(idx[] + 1, 100)
    end
end
```
</details>

[race.webm](https://github.com/MakieOrg/Makie.jl/assets/10947937/254b62e5-200e-4411-9bf4-11d5d5cdf386)

ffreyer commented:
Did a little bit of performance testing: (first point with verbose lighting has the same setup as fast lighting)
![Screenshot from 2023-09-29 14-26-43](https://github.com/MakieOrg/Makie.jl/assets/10947937/ba486f62-b38f-43b5-a907-57e7720d7b78)
That seems like pretty much the expected scaling to me. A bit more up-front cost from having a more complex shader and then roughly linear scaling with the number of lights. The frames are taken with calls to `colorbuffer` on an invisible screen. (Much less overhead than calling record)

Some questions for review:
- I kept `shading = :fast` (old algorithm) as the default. Do we want to switch to `:verbose`?
    - Switching would be a bit slower but maybe not that much?
    - As far as I understand each *element* of the arrays carrying light data counts as one uniform component. I have 16k of these, but I don't know if this is device dependent. If it is that might cause issues.
    - Switching to verbose by default would mean we can do #3239
- Are the `shading` option names good? (none, fast, verbose)

SimonDanisch commented:
I dont like the names (:verbose, :none, :fast), and I think would prefer an enum, but I guess that's a yak we can shave once this is ready ;)
For defaults, I've come more and more to the conviction that we should go for quality first with optional fastpaths, so I think we should make the good shading the default ;)

SimonDanisch commented:
Ideally, the user doesn't have to specify the shading mode and things "just work" when adding different light configurations.
I think this could be an option as well, no?

ffreyer commented:
The "just works" option would be to use the multi-light algorithm by default.

> GL_MAX_FRAGMENT_UNIFORM_COMPONENTS
>    data returns one value, the maximum number of individual floating-point, integer, or boolean values that can be held in uniform variable storage for a fragment shader. The value must be at least 1024. 

Ok, maybe I was too worried about uniform limits. This pr should be adding 576 components with default settings, which is safely below the minimum. 

> I dont like the names (:verbose, :none, :fast), and I think would prefer an enum, but I guess that's a yak we can shave once this is ready ;)

This is at the point where I'm doing this kind of stuff. Maybe
```julia
@enum begin
    NoShading
    FastShading
    MultiLightShading
end
```
then?

SimonDanisch commented:
>The "just works" option would be to use the multi-light algorithm by default.

No I meant something like:
```julia
lights = [] # NoShading
lights = [one_simple_light] # FastShading
lights = [many_lights...] # MultiLightShading
```
Then you wouldn't need to define the lighting in two places ;)

jkrumbiegel commented:
cool, I should test the directional light with treemaps for example

ffreyer commented:
If we do decide to switch to a DirectionalLight as a default we should also adjust `shading = :fast` (as well as WGLMakie and CairoMakie) to use that.

Personally I don't have a strong opinion on DirectionalLight vs PointLight as a default. I could see having a static light be helpful to tell viewing angles and DirectionalLight giving a bit more of a "illuminated by the sun" kind of feel, but it also allows for the area you look at directly to be dark. So I guess it's a bit less helpful for reading off data (which you probably shouldn't try with 3D plots anyway) while giving a bit more clarity on orientation and motion?

ffreyer commented:
I have no idea how to update the generated javascript files in WGLMakie so that still needs updates. Other than that GLMakie and CairoMakie should be ok, though there are now more test failures because I made the default light static (+ DirectionalLight).

<details>

<summary>
For playing around with defaults
</summary>

```julia
using GLMakie, FileIO, LinearAlgebra, GeometryBasics

begin
    fig = Figure(resolution = (800, 800))

    Label(fig[1, 1:2], "Light Intensities", tellwidth = false)
    ambient = Slider(fig[2, 1], range = 0.0:0.01:1.0, startvalue = 0.55).value
    Label(fig[2, 2], map(x -> "ambient = $x", ambient), width = 100)
    dir_int = Slider(fig[3, 1], range = 0.0:0.01:2.0, startvalue = 1.0).value
    Label(fig[3, 2], map(x -> "directional = $x", dir_int), width = 100)

    Label(fig[1, 3:4], "Light \"Position\"", tellwidth = false)
    phi = Slider(fig[2, 3,], range = range(0, 360, length=361), startvalue = 45).value
    Label(fig[2, 4], map(x -> "phi = $(x)°", phi), width = 100)
    theta = Slider(fig[3, 3], range = range(-90, 90, 181), startvalue = 30).value
    Label(fig[3, 4], map(x -> "theta = $(x)°", theta), width = 100)

    Label(fig[4, 1], "Show Light Ray", tellwidth = false)
    vis = Toggle(fig[4, 2]).active
    Label(fig[4, 3], "Relative to cam", tellwidth = false)
    relative = Toggle(fig[4, 4]).active

    light_dir = Observable(Vec3f(1))

    lights = [
        AmbientLight(map(x -> RGBf(x, x, x), ambient)),
        DirectionalLight(map(x -> RGBf(x, x, x), dir_int), light_dir)
    ]

    earth = FileIO.load(Makie.assetpath("earth.png"))
    sphere = GeometryBasics.mesh(Tesselation(Sphere(Point3f(0), 1f0), 540))

    # Sphere new
    Label(fig[5, 1:2], "New")
    ax = LScene(fig[6, 1:2], show_axis = true, scenekw = (lights = lights, center = false))
    m = mesh!(ax, sphere, color = :white)
    # m = mesh!(ax, sphere, color = :white, specular = Vec3f(0))
    center!(ax.scene)

    onany(phi, theta, relative, ax.scene.camera.view) do phi, theta, rel, _
        cp = cosd(phi); sp = sind(phi)
        ct = cosd(theta); st = sind(theta)
        if rel
            cc = cameracontrols(ax.scene)
            viewdir = normalize(cc.lookat[] - cc.eyeposition[])
            up = normalize(cc.upvector[])
            right = normalize(cross(viewdir, up))
            dir = ct * cp * viewdir - ct * sp * right - st * up
        else
            dir = -Vec3f(ct * cp, ct * sp, st)
        end
        light_dir[] = normalize(dir)
    end

    linesegments!(ax,
        map(dir -> Point3f[Point3f(0), -2dir], light_dir),
        color = :yellow, visible = vis
    )

    # Earth new
    light_dir2 = Observable(Vec3f(1))
    lights2 = [
        AmbientLight(map(x -> RGBf(x, x, x), ambient)),
        DirectionalLight(map(x -> RGBf(x, x, x), dir_int), light_dir2)
    ]
    ax2 = LScene(fig[7, 1:2], show_axis = true, scenekw = (lights = lights2, center = false))
    mesh!(ax2, Sphere(Point3f(0), 1f0), color = earth)
    center!(ax2.scene)
    onany(phi, theta, relative, ax2.scene.camera.view) do phi, theta, rel, _
        cp = cosd(phi); sp = sind(phi)
        ct = cosd(theta); st = sind(theta)
        if rel
            cc = cameracontrols(ax2.scene)
            viewdir = normalize(cc.lookat[] - cc.eyeposition[])
            up = normalize(cc.upvector[])
            right = normalize(cross(viewdir, up))
            dir = ct * cp * viewdir - ct * sp * right - st * up
        else
            dir = -Vec3f(ct * cp, ct * sp, st)
        end
        light_dir2[] = normalize(dir)
    end
    linesegments!(ax2,
        map(dir -> Point3f[Point3f(0), -2dir], light_dir2),
        color = :yellow, visible = vis
    )


    # old sphere
    Label(fig[5, 3:4], "Old")
    pos = Observable(Point3f(0))
    old_lights = [
        AmbientLight(RGBf(0.55, 0.55, 0.55)),
        PointLight(RGBf(1, 1, 1), pos)
    ]
    ax3 = LScene(fig[6, 3:4], show_axis = true, scenekw = (lights = old_lights,))
    m = mesh!(ax3, sphere, color = :white)
    # m = mesh!(ax2, sphere, color = :white, specular = Vec3f(0))
    cc = ax3.scene.camera_controls
    map!(identity, pos, cc.eyeposition)
    update_cam!(ax3.scene, cc.eyeposition[], cc.lookat[])

    # old earth
    pos = Observable(Point3f(0))
    old_lights = [
        AmbientLight(RGBf(0.55, 0.55, 0.55)),
        PointLight(RGBf(1, 1, 1), pos)
    ]
    ax4 = LScene(fig[7, 3:4], show_axis = true, scenekw = (lights = old_lights,))
    mesh!(ax4, Sphere(Point3f(0), 1f0), color = earth)
    cc = ax4.scene.camera_controls
    map!(identity, pos, cc.eyeposition)
    update_cam!(ax4.scene, cc.eyeposition[], cc.lookat[])

    fig
end
```

</details>

jkrumbiegel commented:
very cool visualizer :) something is not quite correct with the specular highlight:

<img width="351" alt="grafik" src="https://github.com/MakieOrg/Makie.jl/assets/22495855/e03cce42-8ada-4a59-b324-fcd91a7b55b1">

see that the specular highlight stays in front even if the directional light moves to the side? this causes a visible cutoff

jkrumbiegel commented:
Something like this looks much better to me than the current state. And I don't find it disorienting to rotate the brain ;) the sphere looks a bit weird of course

<img width="798" alt="grafik" src="https://github.com/MakieOrg/Makie.jl/assets/22495855/06eef3dd-8d99-4477-be1b-fa4653817737">

jkrumbiegel commented:
@ffreyer I still don't think it's right, shouldn't the specular reflection move with the light? now only the intensity changes. not sure about this, especially with a directional light

ffreyer commented:
If you lock the light position to the camera it should not move, because the light-to-vertex and camera-to-vertex directions have a constant angle between them. If you keep the light direction constant it should and for me it does.

ffreyer commented:
I added some smoothing to the `diff_coeff` because the edge that appears where the light direction and normals become perpendicular annoys me.
![no_smooth](https://github.com/MakieOrg/Makie.jl/assets/10947937/09c97632-ce31-4daa-872c-5808d40fc5af)

This comes from `diff_coeff ~ dot(lightdir, normal) = cos(angle)`, which is a linear function near 90° that ends abruptly as it hits 0. The smoothing function I added smooths that part out: (with an exponent as a parameter for how weak the smoothing is)
![Screenshot from 2023-10-04 17-17-29](https://github.com/MakieOrg/Makie.jl/assets/10947937/44eeb394-8b76-49b6-99f7-066a4fcc40a3)

| No smoothing | Exponent 16 | Exponent 8 (current) |
| --- | --- | --- |
| ![no_smooth](https://github.com/MakieOrg/Makie.jl/assets/10947937/09c97632-ce31-4daa-872c-5808d40fc5af) | ![pow16_smooth](https://github.com/MakieOrg/Makie.jl/assets/10947937/de2a1b8d-f6cd-42ee-9ccc-718c2df56d20) | ![pow8_smooth](https://github.com/MakieOrg/Makie.jl/assets/10947937/57124c26-c44f-461c-9ff3-759f2937f198) |

jkrumbiegel commented:
Nice now it looks quite good, I restarted the session, maybe the change with the specular highlight was only picked up after. Now it moves as I expect. This setting looks somewhat nice:

<img width="799" alt="grafik" src="https://github.com/MakieOrg/Makie.jl/assets/22495855/4756cc21-c111-4146-afd8-e655c6633cf0">

ffreyer commented:
Yea changes to shaders require a `GLMakie.closeall()` or session restart to apply.

I'm wondering as well if we should set `diffuse = Vec3f(1)` by default. Currently it's 0.4, which makes the brightness values kind of magical. E.g. 0.33 + 1.15 (for direct light) is somehow not full white, because it's actually 0.33 + 0.4 * 1.15 = 0.79. With `diffuse = 1` and `specular = 0.5` (scaled by the same factor from 0.2) you'd get the same result with `ambient = 0.33` and `directional_intensity = 0.46`, but now you can just add the intensities to get the maximum diffuse brightness.

jkrumbiegel commented:
Yeah I was wondering about that as well. I think not limiting diffuse by default sounds very reasonable and easier to predict.

ffreyer commented:
Alright, I made those changes but nudged specular down a little and the light intensities up a little.
![Screenshot from 2023-10-04 18-57-35](https://github.com/MakieOrg/Makie.jl/assets/10947937/a03fe5e1-4f98-4bb7-9d9a-9aefbde97e48)

lazarusA commented:
I cannot wait for this to be merged 🤯 , what's the plan? The TODO list is almos done 😄 .

ffreyer commented:
Yea, the only thing that's missing is some WGLMakie updates for the switch from PointLight to DirectionalLight. Since this includes a couple of breaking changes it would go into the 0.20 beta branch. I'm not sure what the exact state of that branch is. I think we all want to get it finished and merged but there are at least the WGLMakie ci failures that need fixing.

ffreyer commented:
@SimonDanisch I moved the calculation of camera relative light directions to javascript. It seems to work fine to me, but it would be good if you took a closer look. I'm pretty unsure of what I'm doing there and just kind of tried till it worked.

---

In general I'm still a bit unsure about defaults. Quite a few refimgs look darker now, probably mostly because of the angled light direction. In some cases that's helpful, I think, because it helps accentuate curvature

| old | new |
| --- | --- |
| ![Screenshot from 2023-10-11 18-35-06](https://github.com/MakieOrg/Makie.jl/assets/10947937/86c75146-9734-4bb8-b905-f8429e3a9d02) | ![Screenshot from 2023-10-11 18-35-12](https://github.com/MakieOrg/Makie.jl/assets/10947937/1d0b49ef-dbbd-4047-ada0-ee494cf590e0) |
| ![Screenshot from 2023-10-11 18-41-57](https://github.com/MakieOrg/Makie.jl/assets/10947937/57ff4a4e-edfa-4b27-a0cb-805565858635) | ![Screenshot from 2023-10-11 18-42-00](https://github.com/MakieOrg/Makie.jl/assets/10947937/807357db-2e59-4e3c-9ebe-b12989a73197) |
| ![Screenshot from 2023-10-11 18-51-17](https://github.com/MakieOrg/Makie.jl/assets/10947937/11cad5d9-03f8-467a-a1c2-95e14847e700) | ![Screenshot from 2023-10-11 18-51-20](https://github.com/MakieOrg/Makie.jl/assets/10947937/2e003bd4-1dd1-49c5-b4c0-cccfc7e0d343) |
| ![Screenshot from 2023-10-11 18-49-12](https://github.com/MakieOrg/Makie.jl/assets/10947937/9a5cfa18-eee2-4ebc-a15c-7dc9ff6e3a6f) | ![Screenshot from 2023-10-11 18-49-15](https://github.com/MakieOrg/Makie.jl/assets/10947937/6313e747-6617-430a-a144-175f5b3b03f5) |

Some others ended up just looking darker or less detailed. For example:

| old | new |
| --- | --- |
| ![Screenshot from 2023-10-11 18-38-57](https://github.com/MakieOrg/Makie.jl/assets/10947937/3b75140c-2572-4702-8844-cce524f50af0) | ![Screenshot from 2023-10-11 18-38-46](https://github.com/MakieOrg/Makie.jl/assets/10947937/0920b78f-d318-4d9e-b921-cbe4d5261d3c) |
| ![Screenshot from 2023-10-11 18-47-00](https://github.com/MakieOrg/Makie.jl/assets/10947937/29261b56-2be9-49e0-85c5-9c0b3a176e61) | ![Screenshot from 2023-10-11 18-47-04](https://github.com/MakieOrg/Makie.jl/assets/10947937/788e5e7b-e0ce-4d8b-aa7d-aa466f466e4b) |
|![Screenshot from 2023-10-11 18-54-18](https://github.com/MakieOrg/Makie.jl/assets/10947937/04a48302-6b6c-4f52-9d77-60017fde3e83) | ![Screenshot from 2023-10-11 18-54-22](https://github.com/MakieOrg/Makie.jl/assets/10947937/bcc5dc27-a663-4ac5-9596-e4ceb9a3b7bc) |

jkrumbiegel commented:
Maybe somethings not right with Axis3. You know how I'm doing this unusual view and projection matrix, maybe that throws off the math so that the light angle is not right anymore?

ffreyer commented:
Hmm, moving scale_matrix to `model` means that is also applies with `space != :data` which is kind of annoying

ffreyer commented on src/interfaces.jl:
```diff
@@ -228,11 +228,14 @@ function connect_plot!(scene::SceneLike, plot::Combined{F}) where {F}
             plot.transformation = t
         end
-        connect!(transformation(scene), transformation(plot))
+        if is_space_compatible(plot, parent)
+            connect!(transformation(parent), transformation(plot))
+        end
```
This only connects transformations of plots and parent plots/scenes if they have compatible spaces. For plots this checks `get(plot, :space, :data)` and for scenes it will always assume `:data` space to be compatible, with for example `:pixel` space in addition with a pixel camera.

jkrumbiegel commented:
With the reversed axes, the lighting direction obviously changes. I don't think that's good in this camera-relative mode. I would say the light should always come from some angle relative to the main camera axis, and if you pass a direction vector manually it should work in data coordinates, so be affected by scaling. Otherwise it's kind of hard to reason about what it means.

ffreyer commented:
> With the reversed axes, the lighting direction obviously changes. I don't think that's good in this camera-relative mode. I would say the light should always come from some angle relative to the main camera axis, and if you pass a direction vector manually it should work in data coordinates, so be affected by scaling. Otherwise it's kind of hard to reason about what it means.

With the changes the reversal also moved to the model matrix, so the light direction actually doesn't change. What changes are the normals generated by a surface. I.e. with a z reversal they'll point downwards and not interact with the light coming from above. With `backlight = 1f0` the light will interact with -normal too and you get:

![Screenshot from 2023-10-12 20-55-36](https://github.com/MakieOrg/Makie.jl/assets/10947937/63d5d16c-afdf-4ac1-ae35-733d9da15d60)

We could also make that a default, for surface...

ffreyer commented:
Fixed this (even though it should be unrelated to this pr)
![Screenshot from 2023-10-27 14-45-18](https://github.com/MakieOrg/Makie.jl/assets/10947937/7b31676b-6588-493b-b300-68cf25508cd8)

ffreyer commented:
Seems like this too :shrug: 
![Screenshot from 2023-10-27 14-51-51](https://github.com/MakieOrg/Makie.jl/assets/10947937/57bbb434-992a-458d-af37-cd3de7af7dd9)

There also seem to be some issues with line (and maybe scatter) depth order:
![Screenshot from 2023-10-27 14-44-42](https://github.com/MakieOrg/Makie.jl/assets/10947937/e171e49b-0759-493a-bed6-03929ffe0241)

ffreyer commented:
As far as I can tell the only issue left here that isn't also on #3113 is with "contour labels 3D". With Axis3 scaling being moved to `model` the text is now being scaled, which makes it tiny. (You can see it with fontsize ~ 300.) The gaps are also too small I think. This requires some changes to `transform_marker`/`scale_primitive` to be fixed, and probably text bounding boxes
